gateway/ln-gateway/src/gateway_module_ng/mod.rs

```diff
@@ -0,0 +1,287 @@
+pub struct GatewayClientContextNG {
+    pub decoder: Decoder,
+    pub tpe_agg_pk: AggregatePublicKey,
+    pub tpe_pks: BTreeMap<PeerId, PublicKeyShare>,
+    pub gateway: Gateway,
+}
+
+impl Context for GatewayClientContextNG {}
+
+impl ClientModule for GatewayClientModuleNG {
+    type Init = GatewayClientInitNG;
+    type Common = LightningModuleTypes;
+    type Backup = NoModuleBackup;
+    type ModuleStateMachineContext = GatewayClientContextNG;
+    type States = GatewayClientStateMachinesNG;
+
+    fn context(&self) -> Self::ModuleStateMachineContext {
+        GatewayClientContextNG {
+            decoder: self.decoder(),
+            tpe_agg_pk: self.cfg.tpe_agg_pk,
+            tpe_pks: self.cfg.tpe_pks.clone(),
+            gateway: self.gateway.clone(),
+        }
+    }
+
+    fn input_amount(
+        &self,
+        input: &<Self::Common as ModuleCommon>::Input,
+    ) -> Option<TransactionItemAmount> {
+        Some(TransactionItemAmount {
+            amount: input.amount,
+            fee: self.cfg.fee_consensus.contract_input,
+        })
+    }
+
+    fn output_amount(
+        &self,
+        output: &<Self::Common as ModuleCommon>::Output,
+    ) -> Option<TransactionItemAmount> {
+        Some(TransactionItemAmount {
+            amount: output.amount(),
+            fee: self.cfg.fee_consensus.contract_output,
+        })
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub enum GatewayClientStateMachinesNG {
+    Send(SendStateMachine),
+    Receive(ReceiveStateMachine),
+}
+
+impl IntoDynInstance for GatewayClientStateMachinesNG {
+    type DynType = DynState;
+
+    fn into_dyn(self, instance_id: ModuleInstanceId) -> Self::DynType {
+        DynState::from_typed(instance_id, self)
+    }
+}
+
+impl State for GatewayClientStateMachinesNG {
+    type ModuleContext = GatewayClientContextNG;
+
+    fn transitions(
+        &self,
+        context: &Self::ModuleContext,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<Self>> {
+        match self {
+            GatewayClientStateMachinesNG::Receive(state) => {
+                sm_enum_variant_translation!(
+                    state.transitions(context, global_context),
+                    GatewayClientStateMachinesNG::Receive
+                )
+            }
+            GatewayClientStateMachinesNG::Send(state) => {
+                sm_enum_variant_translation!(
+                    state.transitions(context, global_context),
+                    GatewayClientStateMachinesNG::Send
+                )
+            }
+        }
+    }
+
+    fn operation_id(&self) -> OperationId {
+        match self {
+            GatewayClientStateMachinesNG::Receive(state) => state.operation_id(),
+            GatewayClientStateMachinesNG::Send(state) => state.operation_id(),
+        }
+    }
+}
+
+impl GatewayClientModuleNG {
+    pub async fn start_send_state_machine(
```

Comment:
nit: bit of a technical name for the high level function that pays the invoice. Any objections to `pay_bolt11_invoice_v2`?